File: waltid-libraries/protocols/waltid-openid4vc-wallet-recovery-keychain/src/iosMain/kotlin/id/walt/wallet2/recovery/keychain/KeychainIdentityRecovery.kt
@file:OptIn(kotlinx.cinterop.ExperimentalForeignApi::class, kotlinx.cinterop.BetaInteropApi::class)

package id.walt.wallet2.recovery.keychain

import id.walt.wallet2.mobile.identity.IdentityRecoveryData
import id.walt.wallet2.mobile.identity.IdentityRecoveryProvider
import id.walt.wallet2.mobile.identity.RecoveryAvailability
import id.walt.wallet2.mobile.identity.RecoveryProtection
import id.walt.wallet2.mobile.identity.RecoveryReceipt
import id.walt.wallet2.mobile.identity.RecoveryScope
import kotlinx.cinterop.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import platform.CoreFoundation.*
import platform.Foundation.*
import platform.Security.*

/** Only accessibility classes compatible with synchronizable Keychain items are offered. */
public enum class SynchronizableKeychainAccessibility {
    /** Recovery material is available while the device is unlocked. */
    WhenUnlocked,
    /** Also permits background access after the first unlock following a restart. */
    AfterFirstUnlock,
}

/**
 * Opt-in iOS recovery provider using synchronizable generic-password items, never Enclave keys.
 * Keychain accepts local writes even when iCloud Keychain is disabled. Public APIs do not expose
 * per-item cloud delivery or deletion completion; this provider reports local acceptance only.
 */
public class KeychainIdentityRecovery(
    /** Stable namespace shared by the app's installations; it must not be a random device identifier. */
    private val namespace: String,
    /** Optional entitled Keychain access group, used consistently for read/write/delete. */
    private val accessGroup: String? = null,
    /** Device-only accessibility classes cannot be selected for this synchronizable record. */
    private val accessibility: SynchronizableKeychainAccessibility = SynchronizableKeychainAccessibility.WhenUnlocked,
) : IdentityRecoveryProvider {
    init {
        require(namespace.matches(Regex("[A-Za-z0-9._-]{1,64}"))) { "Invalid recovery namespace" }
        require(accessGroup == null || accessGroup.isNotBlank()) { "Keychain access group cannot be blank" }
    }
    override val id: String = "keychain:$namespace"
    override val displayName: String = "iCloud Keychain recovery"
    private val service = "id.walt.wallet.identity.recovery.$namespace"
    private val mutex = Mutex()

    override suspend fun availability(): RecoveryAvailability = withContext(Dispatchers.Default) {
        query().use { query ->
            query.put(kSecUseAuthenticationUI, kSecUseAuthenticationUIFail)
            when (val status = SecItemCopyMatching(query.ref, null)) {
                errSecSuccess, errSecItemNotFound -> RecoveryAvailability.Available(
                    RecoveryProtection.OperatingSystemEndToEnd, RecoveryScope.Cloud)
                errSecInteractionNotAllowed -> RecoveryAvailability.Unavailable("Unlock the device to access Keychain recovery")
                else -> RecoveryAvailability.Unavailable("Keychain recovery is unavailable (status $status)")
            }
        }
    }

    override suspend fun list(): List<String> = mutex.withLock { withContext(Dispatchers.Default) {
        query().use { query ->
            query.put(kSecReturnAttributes, kCFBooleanTrue)
            query.put(kSecMatchLimit, kSecMatchLimitAll)
            memScoped {
                val result = alloc<CFTypeRefVar>(); result.value = null
                when (val status = SecItemCopyMatching(query.ref, result.ptr)) {
                    errSecItemNotFound -> emptyList()
                    errSecSuccess -> {
                        val array = requireNotNull(result.value) as CFArrayRef
                        try {
                            (0 until CFArrayGetCount(array)).map { index ->
                                val item = requireNotNull(CFArrayGetValueAtIndex(array, index)) as CFDictionaryRef
                                val account = requireNotNull(CFDictionaryGetValue(item, kSecAttrAccount))
                                CFBridgingRelease(CFRetain(account)) as String
                            }
                        } finally { CFRelease(array) }
                    }
                    else -> error("Keychain list failed with status $status")
                }
            }
        }
    } }

    override suspend fun store(recordId: String, record: IdentityRecoveryData): RecoveryReceipt = mutex.withLock {
        withContext(Dispatchers.Default) {
            val bytes = record.copyBytes()
            try {
                query(recordId).use { query ->
                    query.put(kSecAttrAccessible, when (accessibility) {
                        SynchronizableKeychainAccessibility.WhenUnlocked -> kSecAttrAccessibleWhenUnlocked
                        SynchronizableKeychainAccessibility.AfterFirstUnlock -> kSecAttrAccessibleAfterFirstUnlock
                    })
                    query.putRetained(kSecValueData, bytes.toNSData())
                    when (val status = SecItemAdd(query.ref, null)) {
                        errSecSuccess -> Unit
                        errSecDuplicateItem -> {
                            val existing = requireNotNull(read(recordId)) { "Existing recovery record is unavailable" }
                            try { check(existing.contentEquals(bytes)) { "A different recovery record already uses this ID" } }
                            finally { existing.fill(0) }
                        }
                        else -> error("Keychain store failed with status $status")
                    }
                }
                RecoveryReceipt.AcceptedLocally
            } finally { bytes.fill(0) }
        }
    }

    override suspend fun retrieve(recordId: String): IdentityRecoveryData? = mutex.withLock {
        withContext(Dispatchers.Default) {
            val bytes = read(recordId) ?: return@withContext null
            try { IdentityRecoveryData(bytes) } finally { bytes.fill(0) }
        }
    }

    override suspend fun delete(recordId: String): RecoveryReceipt = mutex.withLock { withContext(Dispatchers.Default) {
        query(recordId).use { query ->
            val status = SecItemDelete(query.ref)
            check(status == errSecSuccess || status == errSecItemNotFound) { "Keychain delete failed with status $status" }
        }
        RecoveryReceipt.AcceptedLocally
    } }

    private fun read(recordId: String): ByteArray? = query(recordId).use { query ->
        query.put(kSecReturnData, kCFBooleanTrue)
        query.put(kSecMatchLimit, kSecMatchLimitOne)
        memScoped {
            val result = alloc<CFTypeRefVar>(); result.value = null
            when (val status = SecItemCopyMatching(query.ref, result.ptr)) {
                errSecItemNotFound -> null
                errSecSuccess -> {
                    val data = requireNotNull(result.value) as CFDataRef
                    try {
                        val size = CFDataGetLength(data).toInt()
                        require(size in 1..IdentityRecoveryData.MAX_BYTES) { "Invalid recovery record size" }
                        requireNotNull(CFDataGetBytePtr(data)).readBytes(size)
                    } finally { CFRelease(data) }
                }
                else -> error("Keychain read failed with status $status")
            }
        }
    }

    private fun query(recordId: String? = null): Dictionary = Dictionary().apply {
        put(kSecClass, kSecClassGenericPassword)
        putRetained(kSecAttrService, service)
        put(kSecAttrSynchronizable, kCFBooleanTrue)
        recordId?.let {
            require(it.matches(Regex("[A-Za-z0-9._-]{1,128}"))) { "Invalid recovery record ID" }
            putRetained(kSecAttrAccount, it)
        }
        accessGroup?.let { putRetained(kSecAttrAccessGroup, it) }
    }

    private class Dictionary : AutoCloseable {
        val ref = requireNotNull(CFDictionaryCreateMutable(kCFAllocatorDefault, 0,
            kCFTypeDictionaryKeyCallBacks.ptr, kCFTypeDictionaryValueCallBacks.ptr))
        fun put(key: CFTypeRef?, value: CFTypeRef?) = CFDictionarySetValue(ref, key, value)
        fun putRetained(key: CFTypeRef?, value: Any) {
            val retained = requireNotNull(CFBridgingRetain(value))
            try { put(key, retained) } finally { CFRelease(retained) }
        }
        override fun close() = CFRelease(ref)
    }
    private fun ByteArray.toNSData(): NSData = usePinned { NSData.create(bytes = it.addressOf(0), length = size.toULong()) }
}

File: waltid-libraries/protocols/waltid-wallet-sdk-ios/Sources/WalletSDKKeychainRecovery/KeychainIdentityRecovery.swift
import Foundation
import WalletSDK
#if os(iOS)
import Security

/// Only accessibility classes compatible with synchronizable Keychain records.
public enum SynchronizableKeychainAccessibility: Sendable {
    case whenUnlocked, afterFirstUnlock
}

/// Optional iOS recovery adapter. Local Keychain acceptance does not prove iCloud delivery;
/// cloud synchronization still depends on the user's iCloud Keychain configuration.
public actor KeychainIdentityRecovery: IdentityRecoveryProvider {
    public nonisolated let id: String
    public nonisolated let displayName = "iCloud Keychain recovery"
    private let service: String
    private let accessGroup: String?
    private let accessibility: SynchronizableKeychainAccessibility

    public init(namespace: String, accessGroup: String? = nil,
                accessibility: SynchronizableKeychainAccessibility = .whenUnlocked) {
        precondition(namespace.range(of: "^[A-Za-z0-9._-]{1,64}$", options: .regularExpression) != nil)
        precondition(accessGroup == nil || accessGroup?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false)
        self.id = "keychain:\(namespace)"
        self.service = "id.walt.wallet.identity.recovery.\(namespace)"
        self.accessGroup = accessGroup
        self.accessibility = accessibility
    }

    public func availability() async throws -> WalletRecoveryAvailability {
        var request = query()
        request[kSecUseAuthenticationUI as String] = kSecUseAuthenticationUIFail
        switch SecItemCopyMatching(request as CFDictionary, nil) {
        case errSecSuccess, errSecItemNotFound:
            return .available(protection: .operatingSystemEndToEnd, scope: .cloud)
        case errSecInteractionNotAllowed: return .unavailable(reason: "Unlock the device to access Keychain recovery")
        default: return .unavailable(reason: "Keychain recovery is unavailable")
        }
    }

    public func list() async throws -> [String] {
        var request = query()
        request[kSecMatchLimit as String] = kSecMatchLimitAll
        request[kSecReturnAttributes as String] = true
        var result: CFTypeRef?
        let status = SecItemCopyMatching(request as CFDictionary, &result)
        if status == errSecItemNotFound { return [] }
        try check(status)
        guard let records = result as? [[String: Any]] else { throw IdentityProviderError.rejected }
        return try records.map {
            guard let account = $0[kSecAttrAccount as String] as? String else { throw IdentityProviderError.rejected }
            return account
        }
    }

    public func store(recordID: String, data: Data) async throws -> WalletRecoveryReceipt {
        guard (1...4096).contains(data.count) else { throw IdentityProviderError.rejected }
        var request = try query(recordID: recordID)
        request[kSecAttrAccessible as String] = accessibility == .whenUnlocked ? kSecAttrAccessibleWhenUnlocked : kSecAttrAccessibleAfterFirstUnlock
        request[kSecValueData as String] = data
        let status = SecItemAdd(request as CFDictionary, nil)
        if status == errSecDuplicateItem {
            guard try read(recordID: recordID) == data else { throw IdentityProviderError.conflict }
        } else { try check(status) }
        return .acceptedLocally
    }

    public func retrieve(recordID: String) async throws -> Data? { try read(recordID: recordID) }

    public func delete(recordID: String) async throws -> WalletRecoveryReceipt {
        let status = SecItemDelete(try query(recordID: recordID) as CFDictionary)
        if status != errSecItemNotFound { try check(status) }
        return .acceptedLocally
    }

    private func read(recordID: String) throws -> Data? {
        var request = try query(recordID: recordID)
        request[kSecReturnData as String] = true
        request[kSecMatchLimit as String] = kSecMatchLimitOne
        var result: CFTypeRef?
        let status = SecItemCopyMatching(request as CFDictionary, &result)
        if status == errSecItemNotFound { return nil }
        try check(status)
        guard let data = result as? Data, (1...4096).contains(data.count) else { throw IdentityProviderError.rejected }
        return data
    }

    private func query() -> [String: Any] {
        var result: [String: Any] = [kSecClass as String: kSecClassGenericPassword,
                                    kSecAttrService as String: service, kSecAttrSynchronizable as String: true]
        if let accessGroup { result[kSecAttrAccessGroup as String] = accessGroup }
        return result
    }
    private func query(recordID: String) throws -> [String: Any] {
        guard recordID.range(of: "^[A-Za-z0-9._-]{1,128}$", options: .regularExpression) != nil else { throw IdentityProviderError.rejected }
        var result = query()
        result[kSecAttrAccount as String] = recordID
        return result
    }
    private func check(_ status: OSStatus) throws {
        switch status {
        case errSecSuccess: return
        case errSecInteractionNotAllowed, errSecAuthFailed, errSecUserCanceled: throw IdentityProviderError.interactionRequired
        case errSecMissingEntitlement, errSecParam: throw IdentityProviderError.rejected
        case errSecDuplicateItem: throw IdentityProviderError.conflict
        default: throw IdentityProviderError.temporarilyUnavailable
        }
    }
}
#endif
